### PR TITLE
create env template flow

### PR DESCRIPTION
**Pull request by ianedwards**

### CODE REVIEW

File: dashboard/package.json [hunk 1]
```diff
@@ -107,7 +107,7 @@
     "@babel/preset-typescript": "^7.15.0",
     "@ianvs/prettier-plugin-sort-imports": "^4.1.1",
     "@pmmmwh/react-refresh-webpack-plugin": "^0.4.3",
-    "@porter-dev/api-contracts": "^0.2.155",
+    "@porter-dev/api-contracts": "file:../../api-contracts/generated/js",
```
Comment (ferozemohideen) [inline dashboard/package.json:110]: remove

File: dashboard/src/lib/hooks/useTemplateEnvs.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+const listTemplateEnvsResValidator = z.object({
+  name: z.string(),
+  base64_apps: z.string().array().default([]),
+  base64_addons: z.string().array().default([]),
+});
+
+export const useTemplateEnvs = (): TUseTemplateEnvs => {
+  const { currentProject, currentCluster } = useContext(Context);
+
+  const { data: encodedEnvironments = [], status } = useQuery(
```
Comment (ferozemohideen) [inline dashboard/src/lib/hooks/useTemplateEnvs.ts:27]: any reason to not just combine the useQuery and the useMemo?

File: dashboard/src/main/home/cluster-dashboard/preview-environments/v2/setup-app/CreateTemplate.tsx [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+        git_repo_name: "",
+        git_branch: "",
+        porter_yaml_path: "",
+      },
+      deletions: {
+        serviceNames: [],
+        envGroupNames: [],
+        predeploy: [],
+        initialDeploy: [],
+      },
```
Comment (ferozemohideen) [inline dashboard/src/main/home/cluster-dashboard/preview-environments/v2/setup-app/CreateTemplate.tsx:101]: nit: this is the same as the default app for the normal app create flow, consider creating a default const for this and putting it near the app validator so we can reuse it

File: dashboard/src/main/home/cluster-dashboard/preview-environments/v2/setup-app/CreateTemplate.tsx [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+    }
+
+    if (isSubmitSuccessful) {
+      return "success";
+    }
+
+    return "";
+  }, [isSubmitting, errorMessagesDeep]);
+
+  const onSubmit = handleSubmit((data) => {
```
Comment (ferozemohideen) [inline dashboard/src/main/home/cluster-dashboard/preview-environments/v2/setup-app/CreateTemplate.tsx:140]: nit: consider making a form context provider that you can wrap this form around as well as the dashboard, makes it clearer to separate the submit logic from the components imo. i've done it so far for datastores, clusters and addons and it seems to work well

Comment (ferozemohideen) [reply dashboard/src/main/home/cluster-dashboard/preview-environments/v2/setup-app/CreateTemplate.tsx:140]: not as much of a big deal if the app template isn't updatable after creation

**Review by ferozemohideen — APPROVED** [review]

File: api/server/handlers/porter_app/templates_list.go [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+	decoderValidator shared.RequestDecoderValidator,
+	writer shared.ResultWriter,
+) *ListEnvironmentTemplatesHandler {
+	return &ListEnvironmentTemplatesHandler{
+		PorterHandlerReadWriter: handlers.NewDefaultPorterHandler(config, decoderValidator, writer),
+	}
+}
+
+// Environment is a partially encoded environment object
+type Environment struct {
```
[the remark this answers is not in the document]
Comment (ianedwards) [reply api/server/handlers/porter_app/templates_list.go:38]: I'll update

File: dashboard/src/lib/hooks/useTemplateEnvs.ts (hunk 2 above)
Comment (ianedwards) [reply dashboard/src/lib/hooks/useTemplateEnvs.ts:27]: honestly just felt like it was easier to read

File: dashboard/src/main/home/cluster-dashboard/preview-environments/v2/setup-app/CreateTemplate.tsx (hunk 4 above)
Comment (ianedwards) [reply dashboard/src/main/home/cluster-dashboard/preview-environments/v2/setup-app/CreateTemplate.tsx]: done, moved all the form methods into a separate context

**Pull request merged**